### DISCUSSION TITLE
Simple video playback

### DISCUSSION THREAD

**Discussion by Axle-Ozz-i-sofT** [raysan5/raylib#2488] [Ideas]
Hi Ray
I am asking if you already have or are considering a simple video (MMedia) playback library to include in raylib?

I am currently writing a series of books to introduce my fellow students, and and keen beginners, to the world of software programming. I have completed the first two books of five (Final draft) and handed it out to my tertiary lecturer and some fellow ICT students for appraisal and feedback. The focus of the first book is on "Programming" and non language specific. All examples are orientated toward fundamental programming principles rather than being language specific. I have provided examples in C, FreeBASIC and Python 3 for all topics (all examples are cross platform for Win x86-64 and Linux (Ubuntu) x86-64). The reason is that all 3 language examples follow the same flow and logic throughout. A learner can place the example side by side and they read the same routine by routine except for some language specific syntax.
The second book focuses upon setting up the development libraries and IDEs for the 3 languages.

Each of the last 3 books will be language specific, but still follow the same emphasis on keeping the examples to be readable side by side with the same flow and logic. After some extensive research I have decided that your raylib makes for an excellent choice to introduce libraries and some intermediate programming techniques. raylib is one of the few, if not only simple library that is cross platform and available for C, FreeBASIC and Python without me having to spend a small portion of my life writing binders and wrappers. I should be able to integrate raylib and show many fundamental techniques :)
The only thing I am missing in a simple and convenient way is a simple video playback library. I can do this using the language and platform specific toolsets, but no one size fits all at the moment.
For now I am looking at a choice between MPV(lib), VLC(lib) or ffmpeg(lib) but feel like a bit of overkill for a simple playback module. I don't want to introduce too much library and dependency complexity if I can avoid it :) I want to try and encourage a desire to learn programming rather than scare them away lol

My current example list (still in thought):

- Common math and string routines. including basic file I/O and pipes(IPC). **stdlib**
- TUI basic examples using **NCursers** and **PDCursers**.
- Graphics using **raylib** (BGI graphics.h like).
- Data Plotting using either **ralib** or **GNU Plot** (undecided).
- GUI using **raylib**.
- Audio using **raylib**.
- Video (MPV, VLC, ffmpeg ffplay) **Your thoughts?**.
- Database using **Flat File CSV** and **SQLite**
- Practical Task incorporating all of the above examples (Only GUI, no TUI) (Undecided)

I would be appreciative of any thoughts on a **Simple MMedia (Video) playback library** that you may have?
P.S. I could see simple video playback for game intros or scene changes for the game creating enthusiasts :)
Best regards
Axle AU

**Comment by raysan5** [comment]
Hello @Axle-Ozz-i-sofT, the simplest video/audio decoding library I know is [pl_mpeg](https://github.com/phoboslab/pl_mpeg), I used it with raylib and it works great but only supports `MPEG1 Video` and `MP2 Audio`.

Beside that one, I don't know of other simple, small and portable video-decoding libraries.

**Reply by Axle-Ozz-i-sofT** [reply]
Hi Ray
thank you for the response. That looks like a possible option. Most small single file headers will only take RAW AVI which most people will shy away from. I have added it to my research list.
Still working through the possibilities and narrowing in on libMPV at the moment (A fork of MPlayer). C, and Py headers and bindings are available but I would have to build the FB bindings and libmpv.bi
[](https://mpv.io/manual/master/#embedding-into-other-programs-libmpv)
[](https://www.ccoderun.ca/programming/doxygen/mpv/)
It still weighs in a little heavy for my liking so I will definitely look at PL_MPEG :)

ffmpeg and AVIsynth converters are a dime a dozen, or not hard to even script up an all to MPEG1 Video and MP2 Audio converter for pl_mpeg... may even be able to use ffmpeg to convert small videos on the fly to pl_mpeg.

I am over my head in task atm, but will likely jump into this in a month or so. I will let you know what I come up with and how it works in with you raylib :)
P.S. I am looking forward to working up some example with raylib.. My lecturer is 20 years+ industry trained developer and great mentor so I think he will see the value in your libraries. We just completed creating some python training modules together for QLD Australia CertIII students, and the booklets I am working on now are as a compliment if they want to go a bit deeper :)

Thank you very much for your time
Axle

**Comment by GuvaCode** [comment]
Hello. There is a library that uses for ffmpeg from Andreas Stöckel. [https://github.com/astoeckel/acinerella](https://github.com/astoeckel/acinerella)

It's easy to integrate into raylib. Here is my pascal test. [https://github.com/GuvaCode/Raylib-Examples/tree/main/video/acinerella](https://github.com/GuvaCode/Raylib-Examples/tree/main/video/acinerella)
the truth is there is no timing syncronize

**Reply by Axle-Ozz-i-sofT** [reply]
Hi GuvaCode
Thank you for the example. I do have Acinerella as a possible wrapper for ffmpeg. The example will be helpful to get through making a decision on which lib to run with :)
Axle

**Comment by orcmid** [comment]
Great plan Axle.  Will you publish your books in any on-line/multimedia form?

You might also get some help about this on gamedev.tv,   Although they might be concentrated on the use of production-level game-development tool chains, I was led to raylib by one of their courses.

**Reply by Axle-Ozz-i-sofT** [reply]
I do intend on publishing all five booklets under a CC license together when completed. At this time I am releasing them "in house" to my lecturer and students (at their request an interest). This will afford me some genuine feedback and my lecturer is quite generously offering his time to proof and offer suggestions for corrections, so I will give his students first shot at it :)
I will keep your interest noted and drop you a link in the not too distant future.
There is a lot of focus and tons of books that are language specific, but I am yet to see much content that focuses more on the programming fundamentals. I though using C and BASIC, being the 2 common languages that most modern high level languages seam to draw from, would be a good opportunity to show the similarities between languages using Python as the modern abstraction.. Try and reduce some of the scare factor around all of the different languages and complexity. :)
Axle

**Reply by duchainer** [reply]
Context from the future: seems that the in-progress books are now on https://github.com/Axle-Ozz-i-sofT/A-BEGINNERS-GUIDE-TO-PROGRAMMING @orcmid 

Have a great day everyone :D
